src/Storages/MergeTree/MergeTreeSettings.cpp

```diff
@@ -59,15 +59,19 @@ void MergeTreeSettings::loadFromQuery(ASTStorage & storage_def, ContextPtr conte
                 CustomType custom;
                 if (name == "disk")
                 {
+                    ASTPtr value_as_custom_ast = nullptr;
                     if (value.tryGet<CustomType>(custom) && 0 == strcmp(custom.getTypeName(), "AST"))
+                        value_as_custom_ast = dynamic_cast<const FieldFromASTImpl &>(custom.getImpl()).ast;
+
+                    if (value_as_custom_ast && isDiskFunction(value_as_custom_ast))
+                    {
+                        auto disk_name = DiskFomAST::createCustomDisk(value_as_custom_ast, context, is_attach);
+                        LOG_DEBUG(getLogger("MergeTreeSettings"), "Created custom disk {}", disk_name);
+                        value = disk_name;
+                    }
+                    else
                     {
-                        auto ast = dynamic_cast<const FieldFromASTImpl &>(custom.getImpl()).ast;
-                        if (ast && isDiskFunction(ast))
-                        {
-                            auto disk_name = getOrCreateDiskFromDiskAST(ast, context, is_attach);
-                            LOG_TRACE(getLogger("MergeTreeSettings"), "Created custom disk {}", disk_name);
-                            value = disk_name;
-                        }
+                        value = DiskFomAST::getConfigDefinedDisk(value.safeGet<String>(), context);
```

Comment:
This seems redundant, I think exception about unknown disk would be thrown even if we remove this line of code and all this line does it throwing exception (as `getConfigDefineDisk` returns the first argument which is passed there)

Reply:
This is important.
This is for the case:
```
drop table if exists test;
create table test (a Int32) engine = MergeTree() order by tuple()
settings disk=disk(name='02963_custom_disk', type = object_storage, object_storage_type = local_blob_storage, path='./02963_test1/');

drop table if exists test;
create table test (a Int32) engine = MergeTree() order by tuple()
settings disk='02963_custom_disk'; -- { serverError BAD_ARGUMENTS }
```

This case is handled by `getConfigDefinedDisk`, because it is not a disk function.  
`getConfigDefinedDisk` ensures  that the disk is defined by the config and it is not a custom disk.

Reply:
Ok, makes sense

Reply:
Though the method still seems strange. I'd rather make it as `void checkDiskLoadedFromConfig(const std::string & disk_name)` or `void checkDiskIsNotCustom(const std::string & disk_name)`

Reply:
Thnx, I will change it.